misk-testing/src/main/kotlin/misk/service/CachedTestService.kt

```diff
@@ -0,0 +1,71 @@
+package misk.service
+
+import com.google.common.util.concurrent.AbstractIdleService
+import misk.logging.getLogger
+import java.util.concurrent.atomic.AtomicBoolean
+
+/**
+ * Test services can implement CachedTestService
+ * if they'd like to reuse the same service for the span
+ * of a given runtime. This is helpful when you want to avoid
+ * incurring the cost of service startup and shutdown with
+ * each test run.
+ *
+ * NOTE: The caching is only useful if the implementing service
+ * references a shared instance of their underlying resources.
+ * A common way to do this is to leverage a companion object.
+ *
+ * Example:
+ *
+ * @Singleton class TestService : CachedTestService() {
+ *   override fun actualStartup() {
+ *     service.start()
+ *   }
+ *
+ *   override fun actualShutdown() {
+ *     service.stop()
+ *   }
+ *
+ *   companion object {
+ *     private val service = Service()
+ *   }
+ * }
+ *
+ * Sharing the same underlying resources also means that
+ * your tests need to be more conscious about accessing
+ * resources scoped uniquely to that test run, or else
+ * making sure that they cleanup resources before they're run.
+ * This is similar to how we work with DBs, where either
+ * you run something like truncate tables before your tests,
+ * or you ensure that all your test statements are made
+ * relative to the scope of your test.
+ */
+abstract class CachedTestService : AbstractIdleService() {
+  final override fun startUp() {
+      if (hasStarted.compareAndSet(false, true)) {
+        log.info { "starting $javaClass.name" }
+        actualStartup()
+        Runtime.getRuntime().addShutdownHook(Thread {
+          log.info { "stopping $javaClass.name" }
+          actualShutdown()
+        })
+      } else {
+        log.info { "$javaClass.name already running, not starting anything" }
+      }
+  }
+
+  final override fun shutDown() {
+    log.info { "$javaClass.name is cached, will shutdown on runtime shutdown" }
+  }
+
+  companion object {
+    private var hasStarted = AtomicBoolean(false)
```

Comment:
why var?

Reply:
also making it a member of the companion object means it's shared across everything that derives from `CachedTestService` - is that the intent, or is there supposed to be one `hasStarted` per service type?

Reply:
ah right that's not the intent, and this build is going to fail... lemme think about this some more.

Reply:
thinking of having a concurrent map to maintain a bool per service

Reply:
> why var?

ugh I don't write enough kotlin. Changed to val

Reply:
the tests passed, but that's because right now the two usages are across two different modules. Good catch! Pushing fix in a sec.

Reply:
updated and added a test case